Recursive endpoint for retrieving details about an inscription

devords opened the pull request:
This should be useful for a variety of use cases.

elocremarc commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,56 @@ impl Server {
+        genesis_fee: entry.fee,
+        genesis_height: entry.height,
+        timestamp: timestamp(entry.timestamp).timestamp(),
+      })
```
We should add output as well I think so we can eventually do a output recursive endpoint. Output was the first endpoint I tried working on like 4 months ago. This would have tons of uses. You could do dynamic content like a website that you can update based on coin control. Having an inscription aware of what output its on is how you would do this.

elocremarc commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,56 @@ impl Server {
+      Json(InscriptionDetailsJson {
+        address,
+        inscription_number: entry.inscription_number,
+        content_type: inscription.content_type().map(|s| s.to_string()),
```
Can't we just get this from the response of /content?

devords replied:
That'd require loading the content though. If you want to fetch content types for a couple of inscriptions, you probably wouldn't want to load their content too.

elocremarc replied:
True. This comment can probably be resolved.

elocremarc commented:
I was thinking about making this endpoint and I would have done the ones you didn't include haha.
What I have wanted for months are
Address
Sat
Output

We should add output so we can eventually do a output recursive endpoint.  This would have tons of uses. You could do dynamic content like a website that you can update based on coin control. Having an inscription aware of what output its on is how you would do this.

Address would also play into this and all those combined gets us very close to making a PSBT marketplace all on inscriptions.

devords commented:
I wanted to keep the response as minimal as possible, as we can always add stuff. But probably never remove. Happy to add whatever we want, if there's consensus. Output sounds good to me!

Re: sat, I think that'd require sat index, no? That's why I didn't add it.

elocremarc commented:
> Re: sat, I think that'd require sat index, no? That's why I didn't add it.
Yeah but were already crossing that bridge with sat endpoint. So it would return null probably if no sat index.

elocremarc commented:
> if there's consensus. Output sounds good to me!

@rot13maxi and I were talking on the sat endpoint spaces the other day about how we really want to get to the point where we can start making crazy cool inscriptions that get us closer to calling data on the bitcoin node. Been wanting to see if we can get openordex onto an inscription since it came out. Knowing an inscriptions output is invaluable for this kinda stuff.

devords commented:
I'll work on adding `output` to JSON.

Still not convinced about `sat`. My gut feeling is that we should refrain from adding recursive endpoints that rely on a full sat index. The risk is creating two classes of inscriptions, especially as the index grows larger and slower with every block.

We can achieve most of the things with `/children` recursive endpoint that would be made possible with `sat` endpoint. Agree that `sat` endpoint would make it much easier as there are no sit ups involved in figuring out the inscription id. But it also comes at a cost and maybe something we can punt on in favor of `/children`.

owenstrevor commented:
Would be awesome to add the Sat Endpoint. IMO the Sat Endpoint is the most important endpoint for Ordinals right now. With the Sat Endpoint, you can futureproof your collection so that you can access all of the endpoints that come out in the future, and then you can lock down your work by burning the sat. This would probably also give the dev some breathing room from people wanting stuff right away because they can just add the Sat Endpoint and then get it when it comes out.

devords commented:
I'm curious to know what thing is possible with sat endpoint that's not possible with a children endpoint?

devords commented:
@elocremarc Added `satpoint` to response JSON.

elocremarc commented:
> I'm curious to know what thing is possible with sat endpoint that's not possible with a children endpoint?

Sat is more controlled by creator and child can be collector basically. However you could make the sat endpoint work for a collector if you reference the sat that an inscription is being inscribed to. With sat you can know what the sat is prior to inscribing. However if you want to know the sat in some future child its good to have it here in this endpoint because it might not be known ahead of time.

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+
+    let entry = index
+      .get_inscription_entry(inscription_id)?
+      .ok_or_not_found(|| format!("inscription {inscription_id}"))?;
```
We can `.unwrap` here, because we should never have an inscription in the inscription to ID table without a corresponding entry.

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+
+    let satpoint = index
+      .get_inscription_satpoint_by_id(inscription_id)?
+      .ok_or_not_found(|| format!("inscription {inscription_id}"))?;
```
We can unwrap here too.

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+
+    let address = output
+      .as_ref()
+      .and_then(|o| page_config.chain.address_from_script(&o.script_pubkey).ok())
```
```suggestion
      .and_then(|output| page_config.chain.address_from_script(&output.script_pubkey).ok())
```

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+    Ok(
+      Json(InscriptionDetailsJson {
+        address,
+        inscription_number: entry.inscription_number,
```
```suggestion
        number: entry.inscription_number,
```

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+      Json(InscriptionDetailsJson {
+        address,
+        inscription_number: entry.inscription_number,
+        content_type: inscription.content_type().map(|s| s.to_string()),
```
```suggestion
        type: inscription.content_type().map(|s| s.to_string()),
```

devords replied:
@casey @raphjaph "type" is a reserved keyword, so changign this resulted in a bunch of errors. So, I kept `content_type` and `content_length` as I wasn't sure how to change `type`.

raphjaph replied:
content_type should be fine

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+        address,
+        inscription_number: entry.inscription_number,
+        content_type: inscription.content_type().map(|s| s.to_string()),
+        content_length: inscription.content_length(),
```
```suggestion
        content: inscription.content_length(),
```

raphjaph replied:
content_length is fine

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+        inscription_number: entry.inscription_number,
+        content_type: inscription.content_type().map(|s| s.to_string()),
+        content_length: inscription.content_length(),
+        genesis_fee: entry.fee,
```
```suggestion
        fee: entry.fee,
```

casey commented on src/subcommand/server.rs:
```diff
@@ -712,6 +717,57 @@ impl Server {
+        content_type: inscription.content_type().map(|s| s.to_string()),
+        content_length: inscription.content_length(),
+        genesis_fee: entry.fee,
+        genesis_height: entry.height,
```
```suggestion
        height: entry.height,
```

casey commented on src/templates/inscription.rs:
```diff
@@ -19,6 +19,18 @@ pub(crate) struct InscriptionHtml {
 
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct InscriptionDetailsJson {
+  pub address: Option<String>,
```
```suggestion
  pub address: Option<Address>,
```

devords replied:
@casey @raphjaph I couldn't figure this out. Followed the lead of what we already do with `InscriptionJson`. Can you please help?

casey commented on src/templates/inscription.rs:
```diff
@@ -19,6 +19,18 @@ pub(crate) struct InscriptionHtml {
 
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct InscriptionDetailsJson {
+  pub address: Option<String>,
```
These fields should be sorted.

casey reviewed (changes requested):
Almost there! See comments.

devords commented:
@elocremarc @casey I am planning to work on this again this week. Let me know and I can add whatever other keys to JSON.

elocremarc commented:
Been making some onchain inscription viewers I think having parent would also be super helpful here. Then you can traverse up and down the family tree. 
example: if you are on some inscription then you can see the siblings with the /children endpoint. Or find the ancestor.

> Let me know and I can add whatever other keys to JSON.

My vote adding the `sat_number` and `parent` to the JSON

owenstrevor commented:
> @casey don't we wanna add the sat number in this endpoint? This is very useful combined with the child endpoint or doing the sat endpoint on children you might not know the sat number for prior to inscribing like with batch inscriptions.

YES PLEASE. Right now we have to inscribe the sat number itself, either in the individual inscription metadata or in a lookup table afterwards. Both are hacky solutions where truth is what the dev wrote and not the blockchain. Biggest gap for us right now.

devords commented:
@owenstrevor Yeah, planning to work on this today!

raphjaph commented:
Is this ready for review?

devords commented:
@raphjaph Going to working on the requested changes today!

devords commented:
@raphjaph Updated the branch with the suggested changes. Few open questions:

- I think we should make it clear that this recursive endpoint may add new keys in the future. For example, we may want to add `delegate` etc.
- I added `sat` to the JSON. The request will fail with `404` if sat index isn't found. Can you please review that bit?

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
     })
   }
 
+  async fn inscription_details(
```
Maybe call this `inscription_recursive` so it's clear that this function has stricter guarantees than the normal server functions.

devords replied:
Fixed in debb433585e50e3e3f2615204c6984d5e702db9c

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
+    Extension(index): Extension<Arc<Index>>,
+    Path(inscription_id): Path<InscriptionId>,
+  ) -> ServerResult<Response> {
+    if !index.has_sat_index() {
+      return Err(ServerError::NotFound(
+        "this server has no sat index".to_string(),
+      ));
+    }
```
I think the JSON should just have the sat field as an option instead of requiring a sat index for this endpoint.

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
+      .unwrap()
+      .unwrap();
+
+    let sat = entry.sat.unwrap();
```
Make Option

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
+
+    Ok(
+      Json(InscriptionDetailsJson {
+        address,
```
We do not want to return an address, so as to discourage address reuse.

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
+        fee: entry.fee,
+        height: entry.height,
+        number: entry.inscription_number,
+        postage: output.as_ref().map(|o| o.value),
```
```suggestion
        value: output.as_ref().map(|o| o.value),
```

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
+        number: entry.inscription_number,
+        postage: output.as_ref().map(|o| o.value),
+        sat,
+        satpoint,
```
maybe add the output field as well

devords replied:
Added!

raphjaph commented on src/templates/inscription.rs:
```diff
@@ -20,6 +20,20 @@ pub(crate) struct InscriptionHtml {
 }
 
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct InscriptionDetailsJson {
```
```suggestion
pub struct InscriptionRecursiveJson {
```

raphjaph commented on tests/server.rs:
```diff
@@ -281,6 +281,58 @@ fn inscription_metadata() {
+  bitcoin_rpc_server.mine_blocks(1);
+
+  let output = CommandBuilder::new("wallet inscribe --fee-rate 1 --file wizards.txt")
+    .write("wizards.txt", "THEWIZARDSOFORD")
```
```suggestion
    .write("foo.txt", "FOO")
```

raphjaph commented on tests/server.rs:
```diff
@@ -281,6 +281,58 @@ fn inscription_metadata() {
 }
 
+#[test]
+fn recursive_inscription_metadata() {
```
```suggestion
fn recursive_inscription_endpoint() {
```

raphjaph reviewed (changes requested):
Looking good, have a look at my comments

raphjaph commented:
I think we would also want to add things like charms

raphjaph commented on src/subcommand/server.rs:
```diff
@@ -832,6 +837,75 @@ impl Server {
+    Extension(server_config): Extension<Arc<ServerConfig>>,
+    Extension(index): Extension<Arc<Index>>,
+    Path(inscription_id): Path<InscriptionId>,
+  ) -> ServerResult<Response> {
```
Ah ja and wrap this whole thing in a task::block_in_place

devords commented:
@raphjaph I think I addressed all of the comments, except `address` one.

- Added charms and output
- Made sat optional

Still holding out some hope for including `address`. It'll unlock so many use cases! I am curious if there's some way without having to compromise on address reuse. Maybe the address could be hashed?

raphjaph commented:
What use cases does address unlock?

devords commented:
@raphjaph The primary one I'm interested in is using address for randomization seed for generative art, where rarity would be tied to an address. Moving inscription to a new address would generate a new output. But moving it back to the previous address would then generate the output you had before.

Additionally, you can also do soulbound'esq inscriptions where the inscription behaves different when it's held by a specific owner.

devords commented:
@raphjaph Removed the address

ZedZeroth commented:
@devords Thank you for raising this. I've been thinking about such endpoints since mid-2023 and how much this kind of self-data would benefit the inscription ecosystem. I didn't realise that it was already in progress until @gmart7t2 pointed me here. I would like inscriptions be able to reference their own:

1. Inscription ID
2. Sat number
3. UTXO location
4. Address location
5. Metadata

It looks as though some of these are being pushed through soon? Could you clarify which ones are currently looking likely to be included and then I may respond with some arguments for including the others? :)

ZedZeroth commented:
> @raphjaph The primary one I'm interested in is using address for randomization seed for generative art, where rarity would be tied to an address. Moving inscription to a new address would generate a new output. But moving it back to the previous address would then generate the output you had before.
> 
> Additionally, you can also do soulbound'esq inscriptions where the inscription behaves different when it's held by a specific owner.

Yes 100% this!

ZedZeroth commented:
> @raphjaph I think I addressed all of the comments, except `address` one.
> 
>     * Added charms and output
> 
>     * Made sat optional
> 
> 
> Still holding out some hope for including `address`. It'll unlock so many use cases! I am curious if there's some way without having to compromise on address reuse. Maybe the address could be hashed?

Is this just a general "reusing bitcoin addresses is bad" argument, or is there something more specific to ordinals/inscriptions? Thanks

ep150de commented:
I also have Gen art that depends on the art changing based on the address that holds the inscription.Since a majority use wallets like xverse, unisat, etc. Which uses a constant static receiving address, we need to open the possibility to take advantage of this for creative features such as dynamic art based on the address of the holder that owns it.Sent via the Samsung Galaxy Note20 Ultra 5G, an AT&T 5G smartphone
-------- Original message --------From: ZedZeroth ***@***.***> Date: 2/18/24  04:46  (GMT-08:00) To: ordinals/ord ***@***.***> Cc: Subscribed ***@***.***> Subject: Re: [ordinals/ord] Recursive endpoint for retrieving details about an
  inscription (PR #2628) 

@raphjaph The primary one I'm interested in is using address for randomization seed for generative art, where rarity would be tied to an address. Moving inscription to a new address would generate a new output. But moving it back to the previous address would then generate the output you had before.
Additionally, you can also do soulbound'esq inscriptions where the inscription behaves different when it's held by a specific owner.

Yes 100% this!

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

casey reviewed (approved):
Looks good to me!

casey reviewed (approved):
Looks good to me!

devords commented:
This will be a fun one! Thank you @casey and @raphjaph 🙏

ZedZeroth commented:
Thanks for this. I think I am going to map out what inscriptions can currently access, and what they can't, and then push for "full inter-ordinal operability" where possible. It looks like a sat/inscription's current address is the only major one missing though? I have some further ideas but they may be more technically challenging/prohibitive to include. Huge progress with this has been made in the last six months though, it's fantastic for building :)
